src/webgpu/util/texture/texelData.ts

```diff
@@ -26,263 +21,101 @@ export const enum TexelComponent {
 
 export type PerTexelComponent<T> = { [c in TexelComponent]?: T };
 
-const enum TexelWriteType {
-  Sint,
-  Uint,
-  Float,
-}
-
-// Function to convert a value into a texel value. It returns the converted value
-// and the type of the converted value. For example, conversion may convert:
-//  - floats to unsigned normalized integers
-//  - floats to half floats, interpreted as uint16 bits
-type TexelWriteFn = (value: number) => { value: number; type: TexelWriteType };
+export type ComponentDataType = 'uint' | 'sint' | 'unorm' | 'snorm' | 'float' | 'ufloat' | null;
 
-// Converts a data value to its representation in a shader
-type DecodeFn = (value: number) => number;
-
-interface SingleComponentInfo {
-  decode: DecodeFn;
-  write: TexelWriteFn;
+type TexelComponentInfo = PerTexelComponent<{
+  dataType: ComponentDataType;
   bitLength: number;
-}
-
-type TexelComponentInfo = PerTexelComponent<null | SingleComponentInfo>;
-
-const kR = [TexelComponent.R];
-const kRG = [TexelComponent.R, TexelComponent.G];
-const kRGB = [TexelComponent.R, TexelComponent.G, TexelComponent.B];
-const kRGBA = [TexelComponent.R, TexelComponent.G, TexelComponent.B, TexelComponent.A];
-const kBGRA = [TexelComponent.B, TexelComponent.G, TexelComponent.R, TexelComponent.A];
-
-const unorm = (bitLength: number) => (n: number) => ({
-  value: floatAsNormalizedInteger(n, bitLength, false),
-  type: TexelWriteType.Uint,
-});
-
-const snorm = (bitLength: number) => (n: number) => ({
-  value: floatAsNormalizedInteger(n, bitLength, true),
-  type: TexelWriteType.Sint,
-});
-
-const uint = (bitLength: number) => (n: number) => ({
-  value: (assertInIntegerRange(n, bitLength, false), n),
-  type: TexelWriteType.Uint,
-});
-
-const sint = (bitLength: number) => (n: number) => ({
-  value: (assertInIntegerRange(n, bitLength, true), n),
-  type: TexelWriteType.Sint,
-});
-
-const decodeUnorm = (bitLength: number) => (n: number) =>
-  normalizedIntegerAsFloat(n, bitLength, false);
-const decodeSnorm = (bitLength: number) => (n: number) =>
-  normalizedIntegerAsFloat(n, bitLength, true);
-const identity = (n: number) => n;
-
-const unorm2 = { decode: decodeUnorm(2), write: unorm(2), bitLength: 2 };
-const unorm8 = { decode: decodeUnorm(8), write: unorm(8), bitLength: 8 };
-const unorm10 = { decode: decodeUnorm(10), write: unorm(10), bitLength: 10 };
-
-const snorm8 = { decode: decodeSnorm(8), write: snorm(8), bitLength: 8 };
-
-const uint8 = { decode: identity, write: uint(8), bitLength: 8 };
-const uint16 = { decode: identity, write: uint(16), bitLength: 16 };
-const uint32 = { decode: identity, write: uint(32), bitLength: 32 };
+}>;
 
-const sint8 = { decode: identity, write: sint(8), bitLength: 8 };
-const sint16 = { decode: identity, write: sint(16), bitLength: 16 };
-const sint32 = { decode: identity, write: sint(32), bitLength: 32 };
+// A function which maps component values to component values.
```

Comment:
nit: Could use `/** doc comment */` for many of these.

Reply:
Done.